### PR TITLE
fix issue 7184 - parse error on *(x)++

### PR DESCRIPTION
**Pull request by arBmind**
* reduced parser to the specified grammar

see #7281 and #7442 for reference as well

### CODE REVIEW

**Review by andralex — APPROVED** [review]
clean code!

File: src/dmd/parse.d [hunk 1]
```diff
@@ -7914,112 +7914,31 @@ final class Parser(AST) : Lexer
...
+                            // if .identifier
+                            // or .identifier!( ... )
+                            if (token.value == TOK.dot)
+                            {
+                                if (peekNext() != TOK.identifier && peekNext() != TOK.new_)
                                 {
-                                    e = parseUnaryExp();
-                                    e = new AST.CastExp(loc, e, t);
-                                    error("C style cast illegal, use `%s`", e.toChars());
+                                    error("identifier or new keyword expected following `(...)`.");
```
Comment (JinShil) [inline src/dmd/parse.d]: This does not appear to be covered by any test.  Please add coverage.  You can add a test to the fail_compilation directory to verify compile errors.

File: src/dmd/parse.d [hunk 2]
```diff
@@ -7914,112 +7914,31 @@ final class Parser(AST) : Lexer
...
                                 }
-                                return e;
+                                e = new AST.TypeExp(loc, t);
+                                e = parsePostExp(e);
                             }
-                        default:
-                            break;
+                            else
+                            {
+                                e = parseUnaryExp();
```
Comment (JinShil) [inline src/dmd/parse.d]: This block does not appear to be covered by any test.  Please add coverage.  You can add a test to the fail_compilation directory to verify compile errors.

**Review by JinShil — CHANGES_REQUESTED** [review]

File: src/dmd/parse.d (hunk 2 above)
Comment (arBmind) [reply src/dmd/parse.d]: This line comes from line 8015.

**Comment by arBmind** [comment]
To make this clear. I have not added any logic. This patch only removes code. This broke no existing unit test. I added a now test that shows that the wrong behavior was fixed.

If there was logic without a test before there is no test now. These should be added in a different PR.

**Comment by JinShil** [comment]
@arBmind We need a copyright assignment declaration from you.  Please see instructions in the Readme of this repository.

File: src/dmd/parse.d (hunk 2 above)
Comment (JinShil) [reply src/dmd/parse.d]: I added a test for this at https://github.com/dlang/dmd/pull/8256.  It should be pulled before this PR to ensure there isn't any breakage.

**Comment by arBmind** [comment]
@JinShil Sorry this is legally not possible in Germany. The law here is basically "the Copyright is always by the author". I can only give away all kinds of "usage rights". If this is not compatible with this project, feel free to close this PR and use my work (the diff) on your own behalf or ignore it entirely.

If you want to be compatible with every law, take a look at the CLA of Microsoft. https://cla.microsoft.com - The only CLA I was able to sign so far.

**Comment by WalterBright** [comment]
If you prefer, it works if you take the Microsoft CLA, s/Microsoft/D Language Foundation/, sign it and email it to us. Thanks!
(The license in the CLA allows us to make those changes!)

**Comment by arBmind** [comment]
@WalterBright the email was sent.

File: src/dmd/parse.d (hunk 2 above)
Comment (arBmind) [reply src/dmd/parse.d]: This new test is now broken by this PR.

**Pull request closed without merging**